flow/damage_context.h

```diff
@@ -0,0 +1,80 @@
+#ifndef FLUTTER_FLOW_CONTEXT_H_
+#define FLUTTER_FLOW_CONTEXT_H_
+
+#include <unordered_set>
+#include "flutter/flow/embedded_views.h"
+#include "third_party/skia/include/core/SkRect.h"
+
+class SkImageFilter;
+
+namespace flutter {
+
+struct PrerollContext;
+class Layer;
+
+class DamageContext {
+ public:
+  typedef bool (*LayerComparator)(const Layer* l1, const Layer* l2);
+
+  // Opaque representation of a frame contents
+  class FrameDescription;
+
+  void InitFrame(const SkISize& frame_size,
+                 const FrameDescription* previous_frame_description);
+
+  void PushLayerEntry(const Layer* layer,
+                      LayerComparator comparator,
+                      const SkMatrix& matrix,
+                      const PrerollContext& preroll_context,
+                      size_t index = size_t(-1));
+
+  size_t layer_entries_count() const { return layer_entries_.size(); }
+
+  // Adjusts the paint bounds of layer entries within given range according to
+  // filter bounds
+  bool ApplyImageFilter(size_t from,
+                        size_t count,
+                        const SkImageFilter* filter,
+                        const SkMatrix& matrix,
+                        const SkRect& bounds);
+
+  struct DamageResult {
+    std::unique_ptr<FrameDescription> frame_description;
+    SkRect damage_rect;
```

Comment:
Rather than hard code this to have a single exposed damage rect, why not have the rect be private and have an add method that can choose to do a single rect or multiple rects. For this first pass it might just join to a single global rect. In the future we can maintain a list of rects.

Readback should support iterating a list of damage rects (which would be just one for now).
Optionally, a global damage rect could be requested for cases where the caller wants to do a quick rejection of something on the entire set of damage regardless of whether the implementation kept multiple rects or not.

Reply:
I wanted to talk about this. I agree that api should expose a list.

In theory we can use SkRegion to get a perfect list of non overlapping rectangles, but the issue here is that we also need to restrict rendering to those rectangles. That'd need either a preroll (for culling) and a paint for each rectangle, or cull only once for union and have skia clip toa  complex shape. 

So we might want to find a middle ground, i.e. if we have two small non overlapping rectangle in each corner, render them separately, but for most overlapped rectangles just join them.

Reply:
I don't think we need to implement the multi-rect accumulation in the first pass, I was just saying that we should create an implementation that will work when we do. SkRegion supports iterating rectangles and if we had an IRect vector that would as well.

We should be able to pre-roll only once and then do multiple paint passes.

Reply:
[This commit](https://github.com/knopp/engine/commit/970713db8ac27ff0c35d23aa99f6354ecdfc9a85#diff-474dbfcd032faa5c3101bb02c6e78ac2R15) introduces `DamageArea` class instead of exposing single rect. The `CompositorContext` behavior is changed so that it culls on `DamageArea::bounds()` and then clips and paints individual rects (`DamageArea::GetRects()`). At this point `DamageArea ` only provides one rect identical to bounds.